src/sensai/util/metadata.py

```diff
@@ -0,0 +1,76 @@
+from typing import Callable, Any
+
+import pandas as pd
+
+
+class DataFrameHistoryTracker:
+    """
+    A simple class for keeping track of changes in data frame metadata (like column names, and indices).
+    The idea is to call update on the result of methods that transform data frames
+
+    Example:
+
+    >>> from sensai.util.metadata import trackDFHistory, DataFrameHistoryTracker
+    >>> import pandas as pd
+
+    >>> df = pd.DataFrame({"bar": [1, 2]})
+    >>> dfHistory = DataFrameHistoryTracker(df)
+    >>> df["foo"] = [4, 5]
+    >>> df.index = ["first", "second"]
+    >>> dfHistory.update(df)
+    >>> dfHistory.indexHistory
+    [RangeIndex(start=0, stop=2, step=1), Index(['first', 'second'], dtype='object')]
+    >>> dfHistory.columnsHistory
+    [Index(['bar'], dtype='object'), Index(['bar', 'foo'], dtype='object')]
+    """
+    def __init__(self, df: pd.DataFrame):
+        self.numUpdates = -1  # after init this will be 0, so init does not count as update
+        self.columnsHistory = []
```

Comment:
In case we use this in a chain of transformations. Then we can see how stuff was changed in each step

Reply:
Something is wrong with the way it is now. The history will only ever contain two entries, because the object is reset to a new object every time the decorated function is called.
Also, I can't think of a scenario where the same transformer instance would be used more than once in a single chain of transformations.

Reply:
For the decorator you are right but anyway it's gone now. The tracker is more general and I though it could be useful outside of the transformers, e.g. in an interactive setting. Say I have a bunch of feature generators and also do some stuff to a df myself in a python session. Then I can do:

```python
df = loadSomething()
tracker = DataFrameHistoryTracker(df)
doSomething(df)
tracker.update(df)
fgen1.apply(df)
tracker.update(df)
doSomethingElse(df)
tracker.update(df)
```

Reply:
I don't think I would ever use this pattern. If I wanted to find out what's happening, I'd use the debugger rather than add such lines to my code.

Reply:
My suggestion is to do what is actually desired here: track column changes for DFTs.
Don't solve problems you don't have. Keep it simple, clear and to the point.

Reply:
I rewrote the class. I still called it tracker since I couldn't think of a better name for the method taking the transformed DF. If you have a better suggestion, I am happy to change the naming

Reply:
The name is fine, though I would go with the full "DataFrame" prefix; we don't need to save letters here.

Reply:
Done. I think all comments were addressed now and we can merge this